### PR TITLE
Monitor node status updates in kubelet for external cloud provider name/ip changes

### LINKED ISSUE #55633: Master Kubelet TLS verification broken due to external cloud provider node initialization order
Before the CCM started initializing node addresses (IP addresses and Hostnames) asynchronously, the kubelet would read the addresses, and hostnames from the node or from the cloud before creating the TLS serving certs. This ensured that the TLS certs reflected the correct addresses.

Now that the asynchronous model is being followed for node initialization, the kubelet generates certs with a set of addresses that do not reflect the cloud initialized addresses for the node. 

The order of node address initialization -> TLS cert creation should be preserved with the new asynchronous model as well

### PR DESCRIPTION
**Pull request by rrati**
The kubelet doesn't pass certificate information to the kublet server on startup, and instead now relies upon a GetCertificate method.  This also handles node name/ip information changes when using an external cloud provider

Fixes #55633 

```release-note
kubelet: Monitor node status updates in kubelet for external cloud provider name/ip changes
```

@kubernetes/sig-node-bugs @liggitt @wlan0

### CODE REVIEW

**Comment by liggitt** [comment]
this is post-1.9, correct? prioritizing code reviews for freeze

File: pkg/kubelet/certificate/kubelet.go [hunk 1]
```diff
@@ -148,3 +153,115 @@ func NewKubeletClientCertificateManager(certDirectory string, nodeName types.Nod
 	}
 	return m, nil
 }
+
+type CertificateUpdater interface {
+	UpdateCertInfo(nodeAddrs []v1.NodeAddress) error
+	//	GenerateCert()
```
Comment (dixudx) [inline pkg/kubelet/certificate/kubelet.go]: Remove this commented line?

**Comment by rrati** [comment]
@liggitt I believe this was wanted to be fixed for 1.9.  @wlan0 do you know?

File: pkg/kubelet/certificate/kubelet.go [hunk 2]
```diff
@@ -148,3 +153,115 @@ func NewKubeletClientCertificateManager(certDirectory string, nodeName types.Nod
...
+	certMutex sync.RWMutex
+	cert      *tls.Certificate
+}
+
+func NewCertificateUpdater(hostName string, certFile, keyFile string, iplist []net.IP, dnslist []string) (CertificateUpdater, error) {
+	c, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		return nil, err
+	}
+	updater := certificateUpdater{
```
Comment (dixudx) [inline pkg/kubelet/certificate/kubelet.go]: No need to declare a new variable. Directly return it.

File: pkg/kubelet/server/server.go [hunk 3]
```diff
@@ -138,7 +138,7 @@ func ListenAndServeKubeletServer(
 		// Passing empty strings as the cert and key files means no
 		// cert/keys are specified and GetCertificate in the TLSConfig
 		// should be called instead.
-		glog.Fatal(s.ListenAndServeTLS(tlsOptions.CertFile, tlsOptions.KeyFile))
+		glog.Fatal(s.ListenAndServeTLS("", ""))
```
Comment (dixudx) [inline pkg/kubelet/server/server.go]: Why changing this?

**Comment by liggitt** [comment]
> I believe this was wanted to be fixed for 1.9

This is new function, and a significant inversion (desired, but a significant change). Requires careful review and I'm unsure whether people familiar enough with the kubelet startup dependencies will have review bandwidth for 1.9 at this point.

**Comment by wlan0** [comment]
@luxas I remember you mentioned that we need this for 1.9. Is that still the case?

File: pkg/kubelet/server/server.go (hunk 3 above)
Comment (rrati) [reply pkg/kubelet/server/server.go]: This is changed to force the usage of GetCertificates.  If certs are given then GetCertificates will not be called and we would need GetCertificates to be called.

File: pkg/kubelet/certificate/kubelet.go (hunk 1 above)
Comment (rrati) [reply pkg/kubelet/certificate/kubelet.go]: Done

File: pkg/kubelet/certificate/kubelet.go (hunk 2 above)
Comment (rrati) [reply pkg/kubelet/certificate/kubelet.go]: Done

**Review by luxas — COMMENTED** [review]
Unfortunately I kinda agree with @liggitt here, kinda late in the cycle and we didn't get the critical mass of reviewers from sig-auth and sig-node.
Although I could review this and try to push it into v1.9, I recognize the risk as too high vs the gain.

WDYT?

**Comment by rrati** [comment]
@kubernetes/sig-auth-pr-reviews

**Comment by rrati** [comment]
@kubernetes/sig-node-pr-reviews

**Comment by luxas** [comment]
Let's revisit this @liggitt after v1.9 is out

cc @cheftako FYI ^

**Comment by mikedanese** [comment]
Won't this cause two certs (one useless) to be created in rapid succession when a node starts? This change is ok but papers over the issue. Rather than changing the certificate manager, I would prefer if the linked issue was fixed by the kubelet waiting for the node object to be initialized before starting the serving certificate manager.

**Comment by liggitt** [comment]
Agree. The flow needs to be updated. 

1. Client setup
2. Node API object setup
3. Watch Node API
4. Update/generate/request cert
Repeat 3-4

**Comment by rrati** [comment]
@kubernetes/sig-api-machinery-api-reviews

**Comment by rrati** [comment]
@liggitt I thought we didn't want to delay the node startup/initialization?  An initial proposal for this change was to use an initializer on the node obj and have the node wait to initialize until the ccm had removed the initializer.  Is that back on the table or are you suggesting a refactoring of the kubelet such that the the kubelet will not generate a cert (and thus not start serving any secure functions) until the node object has its host/ip information?  That the host/ip information in the node obj can come from the local system (in non-cloud provider configuration) or from the ccm (in cloud provider configs)?

**Comment by liggitt** [comment]
/close 
in favor of https://github.com/kubernetes/kubernetes/pull/65594

**Pull request closed without merging**